src/dbnode/integration/commitlog_bootstrap_with_snapshots_after_restart_test.go

```diff
@@ -0,0 +1,155 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package integration
+
+import (
+	"testing"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/integration/generate"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/dbnode/persist/schema"
+	"github.com/m3db/m3/src/dbnode/retention"
+	xtime "github.com/m3db/m3/src/x/time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestCommitLogBootstrapWithSnapshotsAfterRestart(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow() // Just skip if we're doing a short run
+	}
+
+	// Test setup
+	var (
+		ropts     = retention.NewOptions().SetRetentionPeriod(12 * time.Hour)
+		blockSize = ropts.BlockSize()
+	)
+	ns, err := namespace.NewMetadata(testNamespaces[0], namespace.NewOptions().
+		SetRetentionOptions(ropts).
+		SetColdWritesEnabled(true))
+	require.NoError(t, err)
+	opts := NewTestOptions(t).
+		SetNamespaces([]namespace.Metadata{ns}).
+		// Make tick interval short enough to sleep on.
+		SetTickMinimumInterval(100 * time.Millisecond)
+
+	setup, err := NewTestSetup(t, opts, nil)
+	require.NoError(t, err)
+	defer setup.Close()
+
+	commitLogOpts := setup.StorageOpts().CommitLogOptions().
+		SetFlushInterval(defaultIntegrationTestFlushInterval)
+	setup.SetStorageOpts(setup.StorageOpts().
+		SetCommitLogOptions(commitLogOpts).
+		SetMediatorTickInterval(50 * time.Millisecond))
+
+	log := setup.StorageOpts().InstrumentOptions().Logger()
+	log.Info("commit log bootstrap with snapshots after restart test")
+
+	// Start the server with filesystem bootstrapper
+	require.NoError(t, setup.StartServer())
+	log.Debug("server is now up")
+
+	// Stop the server
+	defer func() {
+		require.NoError(t, setup.StopServer())
+		log.Debug("server is now down")
+	}()
+
+	// Write test data
+	log.Info("writing test data")
+	now := setup.NowFn()().Truncate(blockSize)
+	seriesMaps := make(map[xtime.UnixNano]generate.SeriesBlock)
+	inputData := []generate.BlockConfig{
+		{IDs: []string{"foo", "bar"}, NumPoints: 50, Start: now.Add(-5 * blockSize)},
+		{IDs: []string{"foo", "qux"}, NumPoints: 50, Start: now.Add(-4 * blockSize)},
+		{IDs: []string{"qux", "quux"}, NumPoints: 50, Start: now.Add(-3 * blockSize)},
+		{IDs: []string{"corge", "porgie"}, NumPoints: 50, Start: now.Add(-2 * blockSize)},
+	}
+	for _, input := range inputData {
+		testData := generate.Block(input)
+		seriesMaps[xtime.ToUnixNano(input.Start)] = testData
+		require.NoError(t, setup.WriteBatch(testNamespaces[0], testData))
+	}
+
+	// Sleep to allow snapshotting to occur.
```

Comment:
Hm, I've always found sleeping to be quite flakey, is there no way to look at metrics emitted using a test scope. 

e.g. using `setup.Scope().Snapshot().Counters()`:
https://github.com/m3db/m3/blob/c373017b19293f9ce24164d4757431f9141906d4/src/dbnode/integration/index_single_node_high_concurrency_test.go#L263:L286

Reply:
Also if we don't use metrics to wait until the snapshot is done, I also feel like we probably can't guarantee these files were actually snapshotted? It's kind of a race.

Could we perhaps A) use the metric to work out when snapshot has done, B) stop the server, then use the code you already have below to make sure there are entries in the snapshot files? That will guarantee cold snapshots actually occurred. Otherwise it's *possible* snapshots had not have happened yet, data was in commit logs and actually restored from commit logs not snapshots, hence we wouldn't be testing the lifecycle of cold snapshots. I know that's maybe only some % chance of possibility, but we don't want to accidentally slide into thinking we have this covered and find out later this wasn't actually catching a regression when a later regression occurs.